5368 break on emfile

markrwilliams opened the pull request:
## Remove this paragraph

Please have a look at our developer documentation before submitting your Pull Request.

https://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch

## Contributor Checklist:

* [x] There is an associated ticket in Trac. https://twistedmatrix.com/trac/ticket/5368
* [x] The changes pass minimal style checks (see: https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted )
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles )
* [x] I have updated the automated tests.

markrwilliams committed:
Stub EMFILE fix with a test that asserts it has lost a connection.

This tests the same property that an EMFILE-robust implementation
should achieve across all the relevant reactor providers.  It should
succeed on all builders that are not gated by linter checks.

markrwilliams committed:
Explicitly connect to 127.0.0.1 for Windows' sake.

Windows can't connect to 0.0.0.0:

http://twistedmatrix.com/trac/ticket/1472

This is probably sane behavior.  All tests that use TCP transports now
explicitly connect to 127.0.0.1.

markrwilliams committed:
Close incoming connections on EMFILE.

twisted.internet.tcp.Port.startListening now reserves an extra file
descriptor for use when `accept(2)` fails with EMFILE.  The reserved
file is closed and then `accept(2)` retried, so that the resulting
client socket can use the newly-freed file descriptor; then the client
connection is closed and the reserve file re-opened, so that the cycle
can begin anew.

NB: this approach will fail if another thread opens a file between the
time the reservation file is closed and `accept(2)` is called.

markrwilliams commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +858,246 @@ def getPeer(self):
+        available, enter=_reserved, outputs=[_returnTrue],
+        collector=_nthOutput(0))
+    _reserved.upon(
+        _openFailedWithEMFILE, enter=_unreserved, outputs=[])
```
This relies on Automat's transitioning to `_reserved` before the outputs for `_unreserved` --`replace`--> `_reserved` finish running.  That isn't a public API; see https://github.com/glyph/automat/issues/41.

markrwilliams replied:
See also https://github.com/glyph/automat/issues/28

markrwilliams commented on src/twisted/internet/tcp.py:
```diff
@@ -1038,42 +1285,62 @@ def doRead(self):
             else:
                 self.numberAccepts = self.numberAccepts+20
-        except:
+        except BaseException:
```
This snuck into the diff so `pycodestyle` complained about it.  Maybe it should be `Exception`.

exarkun replied:
Preserving the old behavior in _this_ PR seems like the wiser choice to me.  So stick with `BaseException` (I think that's equivalent to bare-except?).  Of course, if the goal is not to change the behavior, not changing the code might be the _best_ choice... and if pycodestyle is making a mistake here (complaining about this pre-existing code), it's making a mistake and we should ignore it (after filing a bug report for the mistake).

markrwilliams replied:
[Looks like a `pycodestyle` bug](https://github.com/PyCQA/pycodestyle/issues/749).  I'll climb that hill after this one :frowning:

markrwilliams commented on src/twisted/internet/tcp.py:
```diff
@@ -1038,42 +1285,62 @@ def doRead(self):
-                        # such a listener is not considered readable, so
-                        # accept(2) will never be called.  Calling accept(2) on
-                        # such a listener, however, does not return at all.
+                    elif e.args[0] == EMFILE and self._reservedFD.available():
```
Let's see if the builders confirm `ENOBUFS`.

markrwilliams committed:
Add automat to Intersphinx URLs.

markrwilliams committed:
Fix incorrect method name in docstring.

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +919,81 @@ def createTestSocket(test, addressFamily, socketType):
+
+    def exhaust(self):
+        """
+        Open file descriptors until C{EMFILE} is reached.
```
I think it could be handy to log the number of descriptors this allocated to reach the limit.

Maybe also worth noting that even if this method raises an exception, it is important to call `release` to undo any partial efforts.

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/f36b5e91953ca1ea3630d506008de8d3a71444a6

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +919,81 @@ def createTestSocket(test, addressFamily, socketType):
+            try:
+                os.close(fd)
+            except Exception:
+                continue
```
Hmmm.  `OSError`?  `IOError`?  `EnvironmentError`?  Ahh, the choices.  What exceptions did you anticipate here?  What exceptions did you observe?  How about squashing OSError for EBADF and logging anything else?

Also, separately, maybe `_fileDescriptors` should be cleared by this method, too?  State, eh. :/

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/f36b5e91953ca1ea3630d506008de8d3a71444a6

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +919,81 @@ def createTestSocket(test, addressFamily, socketType):
+            open(os.devnull).close()
+
+        exception = self.assertRaises(IOError, failsWithEMFILE)
+        self.assertEqual(exception.errno, EMFILE)
```
If either assertion fails, I think all of the descriptors opened by the exhauster remain open and the test process is totally hosed from this point onward.

Perhaps after `exhaust()` this test should `addCleanup(release)` and there can be a second test that focuses on the release logic (then you can remove the "and" clause from the test method description which is usually a good sign).

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/f36b5e91953ca1ea3630d506008de8d3a71444a6

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -977,6 +1054,40 @@ def test_allNewStyle(self):
+        self.connectToListener(reactor, listeningHost, clientFactory)
+
+        exhauster = _ExhaustsFileDescriptors()
+        reactor.callWhenRunning(exhauster.exhaust)
```
This logic depends on the reactor creating the client socket even if the reactor is not running.  All current reactor implementations seem to do this so it seems like the test can get away with this.  I was imagining this test would use the `socket` module directly for the client side.  However, since this works and is less complicated, it makes sense to do it the way you've written.  At worst, someone will have to do some twiddling at some unspecified future point to account for reactor refactoring (as an aside, the way the *Client* classes work, by initiating async I/O operations as a side effect of `__init__` is terrible; ah, 2001...).

markrwilliams replied:
This behavior is indeed pretty terrible.  My understanding is that it's at least part of the reason we're trying to move towards endpoints, whose `connect` and `listen` methods return `Deferred`s...

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -977,6 +1054,40 @@ def test_allNewStyle(self):
+        self.assertRaises(
+            self.lostConnectionReason,
+            clientFactory.lostReason.raiseException,
+        )
```
Does this test ever terminate (aside from a `runReactor` timeout) if the desired case isn't entered?  Let's say the exhauster somehow fails to provoke the EMFILE condition (who knows why).  The connection will be accepted and I guess it will just sit there  doing nothing for a while, right?  Perhaps the test could use a server protocol which notes that the test expectations have been violated and ends the test immediately.

Also, in this case, I don't think `exhauster.release()` will be called and any allocated descriptors will leak.

markrwilliams replied:
Adressed in: https://github.com/twisted/twisted/pull/996/commits/f5ffa9d2c3e4cb05690d679219342f3d770b8b69

Is this excessive?  It feels excessive, but it also seems good to test something that makes the tests fail sanely.

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -988,6 +1099,24 @@ def getListeningPort(self, reactor, factory, port=0, interface=''):
+
+        @return: The connector
+        """
+        return reactor.connectTCP('127.0.0.1', address.port, factory)
```
Would be nice to have this duplication (`ListenTCPMixin.connectToListener`, `SocketTCPMixin.connectToListener`) factored out.

I suppose this hasn't come up before because the existing tests use the `socket` module instead of Twisted's TCP client API (so they just have a bunch of semi-duplicated `socket`-module code).  That makes me think there may have been some explicit decision not to use Twisted's client APIs to test these Twisted server-side behaviors.

Also, if it makes sense to keep this method, it seems like the `address` parameter is a bit misleading since the connection is always to localhost.

markrwilliams replied:
> I suppose this hasn't come up before because the existing tests use the socket module instead of Twisted's TCP client API (so they just have a bunch of semi-duplicated socket-module code). That makes me think there may have been some explicit decision not to use Twisted's client APIs to test these Twisted server-side behaviors.

[It seems like other tests used `socket`s directly to minimize the system under test](https://github.com/twisted/twisted/blob/de597aa01ea3e0db7ff95f31b0fecd74318fd285/src/twisted/internet/test/test_tcp.py#L1096-L1131).  I'll admit to using `connectTCP` here out of laziness: it allows me to use `MyClientFactory` which conveniently fires a `Deferred` on `connectionLost`.

> Also, if it makes sense to keep this method, it seems like the address parameter is a bit misleading since the connection is always to localhost.

[This is an unfortunate bit of platform specificity](https://github.com/twisted/twisted/pull/996/commits/94113c4cf9709f479e2e8f5d458da72066384487).  The other option i saw was to change `getListeningPort`'s default interface to `'127.0.0.1'`, which would cause `port.getHost().address` to be connectable on Windows. This seems like the right thing to do since I don't think tests should listen on my public IP address anyway, but I didn't want change an implicit interface that other tests might rely on.

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/963e05a2ef24c11b5d78cc385868f0d85602aaac

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -2539,3 +2686,107 @@ def test_resolveIPv6(self):
+class FileDescriptorReservationTests(SynchronousTestCase):
+    """
+    Tests for L{_FileDescriptorReservation}.
+    """
```
This whole suite seems like a great candidate for Hypothesis-based testing.  Guess Twisted doesn't use Hypothesis for anything yet, though ... ?

markrwilliams replied:
[Not yet](https://github.com/twisted/twisted/pull/677).  I'm inclined to address that in a smaller PR...

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -2539,3 +2686,107 @@ def test_resolveIPv6(self):
+        Multiple acquisitions without releases open the reservation
+        file exactly once.
+        """
+        self.assertEqual(len(self.fileObjects), 0)
```
This assertion looks like it snuck in from a (non-existent) `test_initializationDoesntOpenAnything`.

markrwilliams replied:
Sorry - this like the precondition in a Hoare triple.  Alternatively: the test is actually asserting that there's a delta of exactly one, so it should check the starting value.

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -2539,3 +2686,107 @@ def test_resolveIPv6(self):
+        Acquiring the reservation makes it available; releasing it
+        makes it unavailable.
+        """
+        self.assertFalse(self.reserveFD.available())
```
Another assertion for a test about initial conditions?

markrwilliams replied:
[As above](https://github.com/twisted/twisted/pull/996/files/d29f39116cf69586ab814a366acca24f64fddc1a#r181253658).  If repetition doesn't make this pattern defensible I can eliminate it :smile:

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -2539,3 +2686,107 @@ def test_resolveIPv6(self):
+        self.reserveFD.acquire()
+        self.reserveFD.release()
+
+        self.assertFalse(toClose.closed)
```
Not sure how this _could_ ever be closed at this point.  Just being defensive, I guess.

markrwilliams replied:
[As above](https://github.com/twisted/twisted/pull/996/files/d29f39116cf69586ab814a366acca24f64fddc1a#r181253658).

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -921,6 +1166,7 @@ def __init__(self, port, factory, backlog=50, interface='', reactor=None):
             self.addressFamily = socket.AF_INET6
             self._addressType = address.IPv6Address
         self.interface = interface
+        self._reservedFD = _FileDescriptorReservation(lambda: open(os.devnull))
```
It seems unnecessarily consumptive of a limited resource to have one of these reserved FDs _per listening port_.  It seems like it would work just as well to share a single reserve between all ports.

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/89cc5c3e9115eeaf18fbbed101cb847ba7302f0f

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -1038,42 +1285,62 @@ def doRead(self):
+                            "EMFILE encountered;"
+                            " releasing reserved file descriptor.")
+                        self._reservedFD.release()
+                        continue
```
On the last iteration of `for i in range(numAccepts):` this `continue` will jump out of the loop which takes it to the end of `doRead` - the reserve FD will not be re-acquired and the port ends up in a weird state that I don't think it can recover from (wrt EMFILE and the reserve fd).

markrwilliams replied:
Oops!  Thanks for catching this implicit dependency.

I see two ways to address this:

1. Add an additional state variable so that a subsequent `doRead` call can know that the reserve file descriptor should be available;
2. Change [`for i in range(numAttempts)`](https://github.com/twisted/twisted/pull/996/files/d29f39116cf69586ab814a366acca24f64fddc1a#diff-d4edf4dd546357655acc84fff0ab7082L1025) to a while loop and increment `numAttempts` here.

The second option is ugly but maybe less so than the first.

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/024d8fb8326a2e2cefe13f2dd664a61993261269

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +858,246 @@ def getPeer(self):
+        file is unclaimed.
+
+        @param fileObject: The file object to close.
+        """
```
Can you expand on this?  It's not clear to me from this docstring what this method is meant to do.  What does it mean to "replace a file" in this context?

markrwilliams replied:
It really means: tell the caller if the provided file was closed because the reservation was made available by a previous call to `release`.  If it wasn't, then the reservation is still held because `EMFILE` has not occurred.

Naming is hard and I'm bad at it.  Maybe `maybeClose` is a better name?

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/124437a0814e35c6dafddfa21ec2f337960af5e1

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -14,10 +14,15 @@
 import struct
 
+import attr
+import automat
```
`from automat import MethodicalMachine` would more closely match the style of this module, I think.

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/066e1c7a31a9988c7f379b3c204eaa9996d1c78a

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -1038,42 +1285,62 @@ def doRead(self):
+                        self._reservedFD.release()
+                        continue
+                    elif e.args[0] in (
+                            EMFILE, ENOBUFS, ENFILE, ENOMEM, ECONNABORTED):
```
Totally optional: It might be nice to hoist this tuple to module scope, along with the comment (which would then have a bit more horizontal space and be easier to read).

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +858,246 @@ def getPeer(self):
+
+    @return: L{None}
+    """
+    list(outputs)
```
I read through the automat docs (all of them, I think) to try to understand this line but I didn't find anything that explains why this is a useful thing to do.  Did I miss something from the docs?

If I didn't miss anything, please file a doc issue against automat to explain the issue this code is addressing and also add a comment here meanwhile.

markrwilliams replied:
https://github.com/glyph/automat/issues/18 is half the issue here and all of the issue with `nthOutput`.  This collector also ensures that the transition it's applied to returns `None`; by default, Automat inputs return a (possibly empty) list of their output method's return values, which makes no sense for, say, `_FileDescriptorReservation.acquire`.

exarkun replied:
As I understand it, this explains the `return None` (below).  I'm specifically asking about the `list(outputs)` here though.

https://github.com/glyph/automat/issues/30 seems to vaguely suggest that it used to be necessary to consume `outputs` for correct operation but it may no longer be so (but it's not clear to me whether that issue is really fixed or if it was accidentally closed).  A link to _that_ ticket here seems appropriate.

A link to https://github.com/glyph/automat/issues/18 for the `return None` and `_nthOutput` might be good as well.

exarkun replied:
I suppose this and the other automat-related code should be deleted now.

exarkun reviewed (changes requested):
Thanks.  This is a very nice start.  The subject is complex but the code is mostly understandable.  There are some issues with the implementation, I think.  Comments inline.  Happy to discuss if you like.

markrwilliams committed:
Improve _ExhaustsFileDescriptors and its tests.

* Exceptions raised during file descriptor exhaustion close
  intermediate, opened file descriptors;
* _ExhaustsFileDescriptors.count returns the number of opened file
  descriptors so that the tests can confirm the behavior of
  _ExhaustsFileDescriptors.exhaust() and
  _ExhaustsFileDescriptors.release();
* The exhaustion tests always release opened file descriptors via a
  cleanup function;
* Test count, exhaust, explicit release, and implicit release after an
  unexpected exception during exhaustion.
* _ExhaustsFileDescriptors the number of file descriptors that were
  opened to reach the process' limit.

markrwilliams committed:
Windows raises WSAEMFILE only for sockets.

_ExhaustsFileDescriptors reaches the process' file descriptor limit by
opening non-socket file descriptors, so its checks should use
errno.EMFILE.

markrwilliams committed:
Import MethodicalMachine directly to maintain stylistic consistency.

markrwilliams committed:
Refactor test_closePeerOnEMFILE into a separate, tested assertion.

assertPeerClosedOnEMFILE is tested to ensure it stops its reactor and
fails its test if its exhauster doesn't trigger EMFILE.

markrwilliams commented on src/twisted/internet/test/test_tls.py:
```diff
@@ -324,6 +325,26 @@ def useIt(reactor, contextFactory):
+        @return: The connector
+        """
+        return reactor.connectSSL(
+            '127.0.0.1', address.port, factory, self.getClientContext())
```
[As above](https://github.com/twisted/twisted/pull/996/files/d29f39116cf69586ab814a366acca24f64fddc1a#r181253167).  These `connectToListener`s are special cases; the interface should in fact accept whatever address came out of `port.getHost()` so that the UNIX domain socket tests work without modification.  The easiest solution is probably a documentation change.

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/963e05a2ef24c11b5d78cc385868f0d85602aaac

markrwilliams committed:
TCP connectToListener gets its own mixin; document implied host.

markrwilliams commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +858,246 @@ def getPeer(self):
+        """
+        if self._fileDescriptor is None:
+            self._openFailedWithEMFILE()
+            return False
```
This is wrong for two reasons: the file given to `replace` _was_ closed, so [user code can't use it](https://github.com/twisted/twisted/pull/996/files/d29f39116cf69586ab814a366acca24f64fddc1a#diff-d4edf4dd546357655acc84fff0ab7082R1324), and why on earth would something named `_returnTrue` ever return `False`.

markrwilliams replied:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/124437a0814e35c6dafddfa21ec2f337960af5e1

markrwilliams committed:
Ensure accept loop iterates once after releasing the reserved FD.

Without this, the loop will terminate without a socket named `skt`,
and the attempt to close it will fail.  This will place the Port in an
unrecoverable state.  (Thanks to @exarkun for pointing this out.)

markrwilliams committed:
Rename _FileDescriptorReservation.replace to .maybeClose.

This isn't a great name but it's a clearer one.  Also, _returnTrue
should always return True because a) the provided socket will have
been closed even if the reserve file descriptor hasn't been reopened
and b) its name says it returns True.

markrwilliams committed:
Use a single reserved file descriptor for all Ports.

This instead of allocating one for each port.  The whole point of this
change is to conserve file descriptors, so it's worth introducing some
global mutable state to spend only one.

markrwilliams reviewed (commented):
[test_closePeerOnEMFILE is a flaky test](
https://buildbot.twistedmatrix.com/builders/ubuntu16.04-py2.7-nodeps/builds/1309/steps/select/logs/stdio):

```
[FAIL]
Traceback (most recent call last):
  File "/buildslave/ubuntu16_04-py2_7-nodeps/Twisted/build/py27-nodeps-nocov-posix/local/lib/python2.7/site-packages/twisted/internet/test/test_tcp.py", line 1349, in test_closePeerOnEMFILE
    connect=self.connectToListener,
  File "/buildslave/ubuntu16_04-py2_7-nodeps/Twisted/build/py27-nodeps-nocov-posix/local/lib/python2.7/site-packages/twisted/internet/test/test_tcp.py", line 1184, in assertPeerClosedOnEMFILE
    noResult, "Server accepted connection; EMFILE not triggered.")
  File "/buildslave/ubuntu16_04-py2_7-nodeps/Twisted/build/py27-nodeps-nocov-posix/local/lib/python2.7/site-packages/twisted/trial/_synctest.py", line 384, in assertFalse
    super(_Assertions, self).assertFalse(condition, msg)
  File "/usr/lib/python2.7/unittest/case.py", line 416, in assertFalse
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: Server accepted connection; EMFILE not triggered.

twisted.internet.test.test_tcp.TCPFDPortTestsBuilder_EPollReactorTests.test_closePeerOnEMFILE
```

I can reproduce this with `trial -u` and other transport tests; strace implies that `epoll` objects are getting garbage collected.  That is, given:

`strace trial -u twisted.internet.test.test_unix.UNIXPortTestsBuilder_EPollReactorTests.test_closePeerOnEMFILE |& tee fail.log`

This appears in the trace:

```
dup(0)                                  = 1017
dup(0)                                  = 1018
dup(0)                                  = 1019
dup(0)                                  = 1020
dup(0)                                  = 1021
dup(0)                                  = 1022
dup(0)                                  = 1023
dup(0)                                  = -1 EMFILE (Too many open files)
close(5)                                = 0
close(6)                                = 0
close(10)                               = 0
close(12)                               = 0
close(13)                               = 0
close(14)                               = 0
close(15)                               = 0
close(16)                               = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=2845, ...}) = 0
write(3, "2018-04-12 17:57:08-0700 [-] EMF"..., 77) = 77
getsockopt(24, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
connect(24, {sa_family=AF_UNIX, sun_path="./tmp2IDgWm.sock"}, 18) = 0
epoll_ctl(17, EPOLL_CTL_ADD, 24, {EPOLLIN, {u32=24, u64=139753940844568}}) = 0
epoll_wait(17, [{EPOLLIN, {u32=23, u64=139753940844567}}], 4, 29977) = 1
accept(23, {sa_family=AF_UNIX}, [110->2]) = 5
```

File descriptors 5, 6, 10, 12, 13, 14, 15, and 16 all represent `epoll` objects, e.g.;

```
...
epoll_create(1024)                      = 15
pipe([16, 17])                          = 0
...
```

The simplest fix I can think of is to put a `gc.collect()` [at the beginning of `test_closePeerOnEMFILE`](https://github.com/twisted/twisted/pull/996/files#diff-f9525083d254dddffa1c74bceca3e8d7R1343).  This seems silly but probably generally the right thing to do, since who knows what files might be lurking. 

`trial -u` with the collection has also successfully completed 10,000 runs without failing.

markrwilliams committed:
Force garbage collection before the file exhaustion test.

The Python runtime occasionally collected epoll objects linger from
other reactor after _ExhaustsFileDescriptors.exhaust triggers EMFILE
but before Port.doRead; as a result, there were unexpectedly available
file descriptors.

Forcing a garbage collection cycle before the assertion provides some
assurance that stale file objects release their descriptors before
_ExhaustsFileDescriptors.exhaust triggers EMFILE and the test's
assumptions remain true.

markrwilliams committed:
Fix linter errors.

Methods with docstrings and no bodies make the linter think that
whatever follows is their body, so they can't have two lines following
them.

markrwilliams committed:
Make _FileDescriptorReservation.replace -> close rename consistent.

markrwilliams committed:
Always force a collection before attempting to reach EMFILE.

As explained in 215af3fc4a6338288fe59bc7da0e9bd31a64275a, uncollected
file objects might be collected after EMFILE is triggered, defeating
the purpose.  Move the gc.collect invocation from that commit into
_ExhaustsFileDescriptors.exhaust so callers don't have to collect the
garbage themselves.

markrwilliams committed:
Clarify _FileDescriptorReservation implementation and API.

Calling an MethodicalMachine input method in an output depends on
undefined behavior.  This commit replaces the public `acquire` input
with a new `reserve` method that uses private inputs to reserve a file
descriptor and transition back into the _unreserved state if that
fails with EMFILE.  The pure `available` transition was removed in
favor of a getter method to minimize the size of the state machine.

markrwilliams committed:
L{} can't reference an @ivar.

Replace it with C{} instead.

markrwilliams committed:
Log the repr of the remote address tuple instead of nothing.

markrwilliams committed:
No reserved file descriptor for Port EMFILE recovery on Windows.

Windows sockets are tracked separately from file objects, with the
result that calling `os.dup(0)` until triggering EMFILE does not cause
`accept` to fail with WSAEMFILE.  Furthermore, triggering WSAEMFILE at
all on Windows requires acquiring many more sockets than on UNIX-like
platforms.

This feature cannot be supported on Windows without a testing strategy
that causes `accept` to fail with WSAEMFILE.  This commit introduces a
null object, _NullFileDescriptorReservation, used in place of
_FileDescriptorReservation on Windows.  EMFILE tests are now also
skipped on Windows.

NB: The EMFILE recovery file descriptor is also a UNIX-only feature in
libuv:

https://github.com/libuv/libuv/blob/v1.x/include/uv-unix.h#L225

markrwilliams committed:
Hoist accept errors and their documentation to module level.

This makes the comment easier to read.

markrwilliams committed:
Increase coverage.

* An additional `accept` is granted when EMFILE is encountered and the
  reserved FD is available;
* _ExhaustsFileDescriptors.exhaust raises non-EMFILE IOErrors and
  OSErrors;
* _ExhaustsFileDescriptors.release continues to close file descriptors
  despite EBADF errors.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +859,359 @@ def getPeer(self):
+    def maybeCloseForReserved(fileObject):
+        """
+        Close the provided file if and only if L{release} has been
+        called to free the reserved file's file descriptor in exchage
```
typoe - `exchage`

markrwilliams commented:
A better citation inside `libuv` for the claim that a reserved file doesn't work on Windows:

https://github.com/libuv/libuv/blob/v1.x/src/win/stream.c#L49-L65

Versus:

https://github.com/libuv/libuv/blob/v1.x/src/unix/stream.c#L540-L552

markrwilliams committed:
Skip new EMFILE test on Windows.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -1038,42 +1418,48 @@ def doRead(self):
+                        # available after this.
+                        if i == (numAccepts - 1):
+                            numAccepts += 1
+                        continue
```
This looks like it fixes the issue I raised last time.  I'm a bit concerned about the fragility of the fix (with respect to future code changes), though.  The test suite doesn't cover the case where this matters (and it would be tricky to make it and make sure it continues to hit the right case).

exarkun reviewed (changes requested):
Thanks.  Couple comments inline.  Ran out of steam before I got to the test suite.

markrwilliams commented:
@exarkun suggested out-of-band that the `accept` loop inside `Port.doRead` be moved into its own function.  This would in turn allow the reserved file descriptor logic to move closer to the `accept` call and focus the incidental complexity, making the code easier to analyze.  His sketched API:

```python

def accept(reserve, port):
    while True:
        try:
            return port.accept()
        except EMFile:
            with reserve:
                skt, addr = port.accept()
                skt.close()
```

(From https://gist.github.com/exarkun/64fe848c5f8267367e5b01d5c10a93eb)

Note that the context manager would allow `reserved` to re-open itself if `accept` fails with `EAGAIN`, which the current code doesn't.

This seems good so I'm going to implement it.

markrwilliams committed:
Add missing vertical space to appease twistedchecker.

markrwilliams committed:
Factor out accept from Port.doRead; eliminate Automat state machine.

@exarkun noted that `Port.doRead` was hard to comprehend, and that
`_IFileDescriptorReservation`'s API was unclear.  Per his
recommendation, both have been addressed by:

* Factoring out the accept(2) retry logic into its own function.  The
  new `_accept` function returns a generator that yields an infinite
  stream of client connections that is limited by its caller.  It also
  encapsulates the new EMFILE retry logic via a recursive invocation
  that's guarded by `_IFileDescriptorReservation`.available().
* Converting `_IFileDescriptorReservation` into a context manager.
  This helps ensure that the reserved file descriptor is re-opened
  even under error conditions.
* Eliminating the Automat state machine backing
  `_FileDescriptorReservation`.  Automat made managing one state
  variable harder to reason about.

markrwilliams commented:
Ask codecov to reconsider.

markrwilliams committed:
Fix invalid references in docstrings.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+    """
+    L{_IFileDescriptorReservation} implementation.
+
+    @param fileFactory: A factory that will be called to reserve a
```
I think this is supposed to be `ivar`.  If it were on `__init__` then it would be `param`.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+    @param fileFactory: A factory that will be called to reserve a
+        file descriptor.
+    @type fileFactory: A L{callable} that accepts no arguments and
+        returns an object with a C{close} method.
```
Does Twisted do typing?  pep 483, 484, 526?  `typing.Callable[[], ICloseable]`?

markrwilliams replied:
Not yet, for better and for worse.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+                if e.errno == EMFILE:
+                    self._log.failure(
+                        "Could not reserve EMFILE recovery file descriptor.")
+                    self._fileDescriptor = None
```
redundant assignment here, I guess, since the whole thing is guarded by `self._fileDescriptor is None`.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+        if self._fileDescriptor is None:
+            try:
+                fileDescriptor = self._fileFactory()
+            except (IOError, OSError) as e:
```
Not an actionable item but I just learned that on Python >=3.3 `IOError is OSError`.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+        """
+        if not self._fileDescriptor:
+            raise RuntimeError(
+                "No file reserved.  Have you called my reserve method?")
```
I think `RuntimeError` is meant for "errors in the runtime".  I think it is commonly misunderstood to mean "an error occurred at runtime" and so this intention may be blurred past the point of relevance now.

But ... there is no built-in `StateError`.  So, eh.  The utility of the exception signature of an `__enter__` is questionable anyway, since exceptions from this method can only be handled in the most awkward of ways.

I do have some vague idea about `reserve` returning a new object which is then guaranteed to have a working `__enter__` method.  That raises new questions about object lifecycle management of that new object though so, despite temptations otherwise, what you have here is probably the right path.

Also, separately ... re-entrancy?  We may be okay right now but I'm not sure.  At the moment, the only caller of `__enter__` is `_accept`.  Inside the context managed by this context manager, this code runs:

```
                    log.msg("EMFILE encountered;"
                            " releasing reserved file descriptor.")
                    clientsToClose = _accept(listener, reservedFD)
                    for clientToClose, closedAddr in clientsToClose:
                        clientToClose.close()
                        log.msg("EMFILE recovery:"
                                " Closed socket from %r"
                                " and reopening reserve file descriptor"
                                % (closedAddr,))
                    # All pending client connections have been closed.
                    # Return control back to the reactor.
                    return
```

`log.msg` invokes arbitrary application code.  Elsewhere, I've suggested these logging calls should actually be outside of the managed block.  We might need to generalize that: _no_ arbitrary application code can be invoked from that managed block.  If arbitrary application code runs, it might steal the file descriptors and that's the end of our EMFILE safety (at least temporarily).  That's not _too_ bad since we've already acknowledged this protection is a best-effort not a guarantee.

At the top of this section, I was worried we would end up trying to call `__enter__` inside the managed block.  I don't _see_ a way for that to happen now, even with arbitrary application code invoked (short of manually instantiating and running a new reactor - which is not supported anyway).  Future reactor implementation changes _could_ change this, perhaps.  They might be somewhat far-fetched though.  The reactor event loop is not re-entrant itself so it's _not_ going to call `Port.doRead` again while `Port.doRead` is already on the stack.  So this might all be worrying for nothing ... but it has a fragile feeling to it.  I'd probably at least annotate the `with reserveFD:` block with a comment about avoiding application code (we know we would _like_ to eventually tell the application when this case has been encountered so having a reminder here for whoever works on that would be nice) and avoiding code that might want to open file descriptors.

markrwilliams replied:
> I think `RuntimeError` is meant for "errors in the runtime". I think it is commonly misunderstood to mean "an error occurred at runtime" and so this intention may be blurred past the point of relevance now.

It's now documented as a catch-all, both on [Python 2](https://docs.python.org/2/library/exceptions.html#RuntimeError) and [Python 3](https://docs.python.org/3/library/exceptions.html#RuntimeError).  When in Rome...

> I do have some vague idea about reserve returning a new object which is then guaranteed to have a working `__enter__` method. That raises new questions about object lifecycle management of that new object though so, despite temptations otherwise, what you have here is probably the right path.

Also, the context this manages is global state and returning a new context manager whose `__enter__` always succeeds would mask that.  It _does_ seem weird, however.

> Also, separately ... re-entrancy? We may be okay right now but I'm not sure. ...
`log.msg` invokes arbitrary application code. Elsewhere, I've suggested these logging calls should actually be outside of the managed block. We might need to generalize that: no arbitrary application code can be invoked from that managed block. If arbitrary application code runs, it might steal the file descriptors and that's the end of our `EMFILE` safety (at least temporarily). That's not too bad since we've already acknowledged this protection is a best-effort not a guarantee.

Allowing the logging system to steal the recovery FD does seem like an unnecessary trap to set, but when I wrote this it seemed more important to report the behaviors as they happened -- probably because I was running manual integration tests.  Now I think it's better to wrap `_accept` in its _own_ context manager that buffers log messages and flushes them in its `__exit__`.  That would lift logging out of this critical section but still ensure that it occurs.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+            self.reserve()
+        except Exception:
+            self._log.failure(
+                "Could not reserve EMFILE recovery file descriptor.")
```
Maybe "re-reserve" in this message?

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+if platformType == 'win32':
+    _reservedFD = _NullFileDescriptorReservation()
+else:
+    _reservedFD = _FileDescriptorReservation(lambda: open(os.devnull))
```
If, by chance, you've learned anything about Win32 behaviors while developing this branch, this might be a reasonable place to record them.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+    @return: A generator that yields C{(socket, addr)} tuples from
+        L{socket.socket.accept}
+    """
+
```
What's with this whitespace?

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+                # returns EAGAIN.
+                with reservedFD:
+                    log.msg("EMFILE encountered;"
+                            " releasing reserved file descriptor.")
```
I would say ... no logging calls inside this `with` block.  Log observers seem like prime candidates to steal file descriptors?  (Of course, they might really appreciate being able to steal one in order to get these messages out ... but maybe that's *their* problem.)

Also, `twisted.logger` instead of `log.msg`?  I'm not up on the current story about the Twisted-internal log migration.  I would _guess_ "write all new code with `twisted.logger` and gradually port nearby `twisted.python.log` code."

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+                        clientToClose.close()
+                        log.msg("EMFILE recovery:"
+                                " Closed socket from %r"
+                                " and reopening reserve file descriptor"
```
The "reopening" part is redundant inside this loop.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -1022,58 +1315,29 @@ def doRead(self):
+            clients = _accept(self.socket, _reservedFD)
+            # Limit the infinite stream of clients with the attempts
+            # iterator
+            for accepted, (skt, addr) in izip(attempts, clients):
```
`izip(attempts, clients)` looks a lot like `enumerate(clients, 1)` fwiw (and avoids punishing Python 2 w/ an extra list allocation per doRead call (who knows if that's significant)).

markrwilliams replied:
True, but it would move the termination condition (`if accepted > numAccepts: break`) inside the for loop, which seemed like a potential source of bugs.

exarkun replied:
`islice(enumerate(clients, 1), numAccepts)`?  Whether _that's_ actually an improvement over what's checked in now or not is not so clear to me though.  But hey who doesn't love iterators?

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -1022,58 +1315,29 @@ def doRead(self):
                 self.numberAccepts = self.numberAccepts+20
-        except:
+            else:
+                self.numberAccepts = accepted
```
So... `if not accepted` - that means "accept() gave us no connections at all" - then increase numberAccepts by 20.  Otherwise ("we got at least one connection from accept()") set numberAccepts to the number of connections we accepted.

I think maybe this isn't what the old code intended.  I _think_ the intent of the old code was "try to accept up to `numberAccepts`; if you accept that many, next time around try accepting up to 20 more; if you accept fewer, try accepting exactly as many as you accepted this time next time".

Or I guess more succinctly:

```
if accepted == numAccepts:
    self.numberAccepts += 20
else:
    self.numberAccepts = max(1, accepted)
```

or at least something close to that.

markrwilliams replied:
> I think maybe this isn't what the old code intended. I think the intent of the old code was "try to accept up to numberAccepts; if you accept that many, next time around try accepting up to 20 more; if you accept fewer, try accepting exactly as many as you accepted this time next time".

Whoops!  On a closer reading of the original code that seems exactly right:

```python
numberAccepts = self.numberAccepts
...
for i in range(numberAccepts):
    ...
        if e.args[0] in (EWOULDBLOCK, EAGAIN):
            self.numberAccepts = i
            break
    ...
else:
    self.numberAccepts = self.numberAccepts+20
```

(Excerpted from [here](https://github.com/twisted/twisted/blob/dffc6ab317e06dc6bc3dadc1003228c593ea3cc1/src/twisted/internet/tcp.py#L1019-L1075).)

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -11,19 +11,25 @@
 import errno
+import gc
+import io
+import os
```
old macdonald had a farm

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +931,367 @@ def createTestSocket(test, addressFamily, socketType):
+    A class that triggers C{EMFILE} by creating as many file
+    descriptors as necessary.
+
+    @param fileDescriptorFactory: A factory that creates a new file
```
`ivar`?

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +931,367 @@ def createTestSocket(test, addressFamily, socketType):
+        """
+        Open file descriptors until C{EMFILE} is reached.
+        """
+        gc.collect()            # Force a collection to close dangling
```
Comments above instead of to the right are preferred I think.

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +931,367 @@ def createTestSocket(test, addressFamily, socketType):
+                    self._fileDescriptors.append(fd)
+        except Exception as e:
+            self.release()
+            raise e
```
Does this produce a better report?

```
except Exception:
    self.release()
    raise
```

I think that it does on Python 2, at least.  Maybe Python 3 exception chaining makes it do something usable as-written?

markrwilliams replied:
It certainly does on Python 2 - this is left over from some other change, I think.

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +931,367 @@ def createTestSocket(test, addressFamily, socketType):
+        """
+        self.exhauster.exhaust()
+        self.exhauster.release()
+        self.openAFile()        # Does not fail with EMFILE
```
Comment above instead of to the right.

exarkun replied:
Maybe a useful variation on this test would be to assert that `count()` returns the same number after one and two `exhaust` / `release` pairs?  Or maybe that would be too fragile (for some reason).

markrwilliams replied:
How would that differ from https://github.com/twisted/twisted/pull/996/files/#diff-f9525083d254dddffa1c74bceca3e8d7R1066 ?

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +931,367 @@ def createTestSocket(test, addressFamily, socketType):
+        fileDescriptors = []
+        def failsAfterThree():
+            if len(fileDescriptors) == 3:
+                raise ValueError
```
`raise ValueError` produces a vexingly untraceable exception.  I suggest putting a distinctive string in here.  For example, `raise ValueError("test_fileDescriptorsReleasedOnFailure fake open exception")`

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +931,367 @@ def createTestSocket(test, addressFamily, socketType):
+
+        for fd in fileDescriptors:
+            exception = self.assertRaises(OSError, os.close, fd)
+            self.assertEqual(exception.errno, errno.EBADF)
```
Kind of risky since those integers technically lose all meaning as descriptors after they get closed by the exhauster...  Probably be okay and at least the assertion will fail if somehow the integers get re-allocated as descriptors in the interim.

Maybe it would be safer to `os.fstat` them instead of `os.close`ing them though.

markrwilliams replied:
Excellent point - the check shouldn't have side effects on process-global state.

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +931,367 @@ def createTestSocket(test, addressFamily, socketType):
+
+    @param testCase: The test case under which to run this assertion.
+    @type testCase: L{trial.unittest.SynchronousTestCase} or
+        L{trial.unittest.TestCase}
```
`trial.unittest.TestCase` is-a `trial.unittest.SynchronousTestCase` fwiw.

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -913,10 +931,367 @@ def createTestSocket(test, addressFamily, socketType):
+    clientFactory = MyClientFactory()
+    connect(reactor, listeningHost, clientFactory)
+
+    reactor.callWhenRunning(exhauster.exhaust)
```
I'm slightly worried that the connection created by `connect` could be handled by the port before `exhauster.exhaust` runs.  I think this _could_ happen if there were a Deferred-returning "before" "startup" trigger in the reactor because `callWhenRunning` calls the given callable in the "after" "startup" phase.

This could easily be dealt with by using `addSystemEventTrigger("before", "startup", ...)` instead of `callWhenRunning`, I think.

This also guarantees the exhaust happens before the reactor starts handling I/O events (which begin only after all "before" "startup" triggers have been run).

markrwilliams replied:
Moving this to `before` `startup` prevents the reactor from creating its signal delivery pipe :frowning_face:  That means this test has an implicit dependency on the internal setup of the reactor.

I guess the alternative is to patch `ReactorBuilder.runReactor` to accept `installSignal` handlers as an optional argument, except that might raise spurious warnings on platforms where `signal` isn't available.  Then again, do we support any of those?

FWIW it appears that reactors don't wait for `startup` events before performing I/O:

https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/base.py#L753
https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/base.py#L699-L704

Note the absence of `Deferred`s coordinating with `doIteration` et. al.

exarkun replied:
> Moving this to before startup prevents the reactor from creating its signal delivery pipe ￼

Woops!  Guess there's too much random unspecified stuff going on in the reactor to do much better than what's here, then.  Maybe someday.

> FWIW it appears that reactors don't wait for startup events before performing I/O:

True but it does _run_ them.  So it seemed like this would be okay - because `exhaust` doesn't return a Deferred, it just does its job and then it's done.  So not waiting would be okay.  But given the other problems here, it's moot.

exarkun commented on src/twisted/internet/tcp.py:
```diff
@@ -853,6 +856,295 @@ def getPeer(self):
+        """
+        See L{_IFileDescriptorReservation.__enter__}.
+        """
+        if not self._fileDescriptor:
```
`if self._fileDescriptor is None` for consistency with other new code and more precise statement of the intent.

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -2539,3 +2966,197 @@ def test_resolveIPv6(self):
+        """
+        with self.assertRaises(RuntimeError):
+            with self.reservedFD:
+                "Does not run."
```
maybe `side_effect = True` w/ an assertion afterwards to be sure (initially thought `assertFalse` but that relies on correct exception propagation behavior which it would be preferable to isolate this test from).

markrwilliams replied:
This was a lame trick to avoid lowering coverage.

exarkun replied:
Ah, bummer. :/  If these were Hypothesis tests then you could very easily avoid the need  for such things. :wink:

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -2539,3 +2966,197 @@ def test_resolveIPv6(self):
+        self.reservedFD.reserve()
+        with self.assertRaises(TestException):
+            with self.reservedFD:
+                raise TestException
```
`raise TestException()`  (raising a class is a historical language mistake remains shockingly unfixed)

exarkun replied:
Is this test missing an `assertTrue(self.reservedFD.available())` at the end?  It seems like it is otherwise I don't see how it actually verifies the behavior in the docstring.

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -2539,3 +2966,197 @@ def test_resolveIPv6(self):
+        called = [False]
+        def failsWithSuppressedExceptionAfterSecondOpen():
+            if called[0]:
+                raise SuppressedException
```
`raise SuppressedException()`

exarkun commented on src/twisted/internet/test/test_tcp.py:
```diff
@@ -2539,3 +2966,197 @@ def test_resolveIPv6(self):
+
+        with self.assertRaises(AllowedException):
+            with reservedFD:
+                raise AllowedException
```
`raise AllowedException()`

exarkun commented on src/twisted/python/_release.py:
```diff
@@ -35,6 +35,7 @@
     u"https://python-hyper.org/h2/en/stable/objects.inv",
     u"https://python-hyper.org/priority/en/stable/objects.inv",
     u"https://zopeinterface.readthedocs.io/en/latest/objects.inv",
+    u"https://automat.readthedocs.io/en/latest/objects.inv",
```
no longer necessary I guess.  OTOH Twisted uses automat elsewhere so it's probably a good thing to have.

exarkun reviewed (changes requested):
Thanks!  This was a lot easier to follow (OTOH it's Monday morning instead of Friday afternoon but I'm sure that has nothing to do with it).

Some comments left inline but nothing that will require major changes, I think.

markrwilliams committed:
Remove Automat outputs; they're dead code.

markrwilliams committed:
Document _IFileDescriptorReservation.fileFactory as an ivar.

markrwilliams committed:
Remove redundant assignment.

markrwilliams committed:
Make file descriptor existence checks consistent.

markrwilliams committed:
_FileDescriptorReservation.__exit__ re-reserves the file descriptor.

This action is distinct from _FileDescriptorReservation.reserve, so
its log message should be distinct.

markrwilliams committed:
Don't log in _accept to protect the reserved file descriptor.

The logging system runs arbitrary code through its observers that
might claim the file descriptor reserved for closing accepted clients.
This commit has `_accept` take a `Logger` instance as its first
argument, and has Port use a new `_BuffersLogs` context manager to
buffer `_accept`'s log events until after it's exited.

This also switches `_accept`'s catch-all error log to
`twisted.logger`, which required a change to
SelectReactorTests._acceptFailureTest.

markrwilliams committed:
Correct and test scaling of consecutive accepts.

Per @exarkun:

..the... intent of the old code was "try to accept up to
numberAccepts; if you accept that many, next time around try accepting
up to 20 more; if you accept fewer, try accepting exactly as many as
you accepted this time next time".

Asserts around `numberAccepts` verified this, so this commit makes the
new code do what the old code did and adds a test to prevent future
regressions.

markrwilliams committed:
Attrs classes get ivars because they have an implicit `__init__`.

markrwilliams committed:
Comments go above, not to the left.

markrwilliams committed:
Instantiate exceptions, use messages, and preserve tracebacks.

markrwilliams committed:
TestCase is SynchronousTestCase.

markrwilliams committed:
Check for file descriptor validity without side effects.

markrwilliams committed:
Explicitly check that the block doesn't run.

markrwilliams committed:
Calm pycodestylediff.

markrwilliams committed:
Test _ExhaustsFileDescriptors error cases.

markrwilliams committed:
Explain why the reserved file descriptor is useless on Windows.

markrwilliams committed:
Don't allow EPERM to loop accept(2) infinitely and test catch-all.

`accept(2)` is documented as failing with EPERM when "firewall rules
forbid connection".  That sounds like an error that's fatal for the
listening port.

But since 76b194f5b9e70883a7247d21102a746ffab21a82, Twisted has
retried `accept(2)` after EPERM failures.  Previously the `_accept`
refactoring assumed that an `accept(2)` that failed with EPERM would
eventually succeed, but there's no documentation, mailing list posts,
or other lore that confirms this.

Now instead of looping infinitely on `accept(2)`, the `_accept`
function receives an iterable that it iterates over to limit the
number of consecutive `accept(2)` calls it makes.  This also enables a
test to be written for its EPERM handling.

Also, test that a `socket.error` whose errno is not handled by
`_accept` is logged.

markrwilliams committed:
Windows has no EPERM and stubs out some errnos.

Skip the EPERM test and don't consider invalid errors on Windows.

exarkun reviewed (approved):
Great.  Thanks.  No coverage report posted yet but the tests look good to me.  Please merge when you're satisfied the test coverage is sufficient.

markrwilliams committed:
Merge branch 'trunk' into 5368-break-on-emfile

markrwilliams committed:
Increase test coverage: don't check if a running reactor is running.

MyClientFactory.deferred and MyClientFactory.failDeferred are fired
via MyClientFactory.clientConnectionLost and
MyClientFactory.clientConnectionFailed, respectively.  Neither of
these methods are run before MyServerFactory.connectionMade is, so
neither can be fired if stopReactorIfServerAccepted has been called.

Remove the unnecessary reactor.running check.

markrwilliams committed:
Increase test coverage: replace sideEffect variable with a string.

The string can't raise an exception, so the only possible source of
the expected `RuntimeError` is _FileDescriptorReservation.__enter__.

markrwilliams committed:
Increase test coverage: replace for/else with a list comprehension.

The assertion is equivalent but the success case doesn't leave an
uncovered branch.

markrwilliams committed:
Increase test coverage: test that too many accepts fails the test.

This test-within-a-test ensures that FakeSocketWithAcceptLimit.accept
does what it claims.
